git/libgit2/repository_client.go

```diff
@@ -0,0 +1,472 @@
+/*
+Copyright 2022 The Flux authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package libgit2
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/ProtonMail/go-crypto/openpgp"
+	git2go "github.com/libgit2/git2go/v33"
+
+	"github.com/ProtonMail/go-crypto/openpgp/packet"
+	"github.com/fluxcd/pkg/git"
+	"github.com/fluxcd/pkg/git/libgit2/transport"
+	"github.com/fluxcd/pkg/gitutil"
+)
+
+// ClientName is the string representation of Client.
+const ClientName = "libgit2"
+
+type Client struct {
+	path       string
+	repository *git2go.Repository
+	remote     *git2go.Remote
+	authOpts   *git.AuthOptions
+	// transportOptsURL is the backbone of how we use our own smart transports
+	// without having to rely on libgit2 callbacks (since they are inflexible
+	// and unstable).
+	// The smart transport contract provided by libgit2 doesn't support any
+	// kind of dependency injection, so it's not possible to pass certain
+	// app level settings like creds, timeout, etc. down to the transport.
+	// transportOptsURL serves as unique key for a particular set of transport
+	// options for a Git repository, which is then used to fetch these options
+	// at the transport level.
+	transportOptsURL string
+}
+
+func NewClient(path string, authOpts *git.AuthOptions) *Client {
+	return &Client{
+		path:     path,
+		authOpts: authOpts,
+	}
+}
+
+var _ git.RepositoryClient = &Client{}
```

Comment:
I wasn't around two weeks ago to emphasize on the naming discussion in https://github.com/fluxcd/pkg/pull/291#discussion_r921514846, but looking at the implementations, it still feel like we don't need `Repository` prefix. `RepositoryClient` interface could just be called `Client` and the other interfaces could just be `Reader` and `Writer`.

Reply:
I think this would be best paired with your other suggestion in that comment, which would make `git/{repostory,libgit2,gogit}` all individual modules.

Reply:
Yes, agreed, that's a good idea. Let's tackle that later in a separate PR? This PR is already very large.